### ISSUE TITLE
Transparent Multicluster: pcluster health checks

### ISSUE THREAD

**Issue by pweil-** [kcp-dev/kcp#285]
Use Case: Transparent Multicluster

Objectives:

- User can add a second location
- Deployed application moves between locations when service is disrupted (cluster down)
- Ingress follows application

Details

KCP should health-check a pcluster to notice when it's unhealthy and its assignments should be scheduled to another pcluster

**Comment by sttts** [comment]
Which of the other tasks does this depend on?

**Comment by jmprusi** [comment]
@sttts the kcp-ingress controller already handles changes on the ingress location.  I guess it is the scheduler's responsibility to switch the Ingress location when a pcluster is down.

**Comment by sttts** [comment]
@jmprusi are you working on this or is it rather @imjasonh's domain?

**Comment by jmprusi** [comment]
@sttts I think @imjasonh is on it, but I can take it if he has no bandwidth for this or prefers to focus on other stuff.

**Comment by sttts** [comment]
Some questions:

1. How do we know that a cluster is healthy or not?
2. For a second location, what is missing? An API type?

**Comment by jmprusi** [comment]
> 1. How do we know that a cluster is healthy or not?

That feels like a "Syncer" job, as we should avoid reaching the pCluster apiserver directly from kcp due to possible networking issues / requiring more permissions. 

The Syncer can report its status (running), adding some extra info like the pCluster Readyz endpoint and updating the proper Cluster Status conditions. With that information, and the last update timestamp, we can infer (minus edge cases :) ) the cluster condition:

- Cluster reported healthy and inside the time threshold? Nothing to do
- Last status report is above a threshold? Trigger a failover
- Cluster reported as unhealthy: Trigger a failover

This is a really basic health check. In the future, we can extend the health conditions to consider capacity calculations to decide whether a cluster is ok or not or other types of metrics.

**Comment by sttts** [comment]
Actually that resembles kubelet heartbeating in the node status. If there is no update, a controller in the control plane will mark the node as unready.

**Comment by eddycharly** [comment]
Would be happy to give it a try, will be a good opportunity to learn more about kcp if it's ok.

**Comment by sttts** [comment]
First step would be an API draft in https://docs.google.com/document/d/1k72fgGBNJzbY1Rkxj9_1WksHZR0cNT5OeadCpWuc-zM/edit#.

**Comment by ncdc** [comment]
This is how cluster-api does a remote cluster apiserver health check:

https://github.com/kubernetes-sigs/cluster-api/blob/18c7ec867f7e8eb690aa7708afe6bbc27fec1498/controllers/remote/cluster_cache.go#L48-L53

https://github.com/kubernetes-sigs/cluster-api/blob/18c7ec867f7e8eb690aa7708afe6bbc27fec1498/controllers/remote/cluster_cache.go#L314-L327

https://github.com/kubernetes-sigs/cluster-api/blob/18c7ec867f7e8eb690aa7708afe6bbc27fec1498/controllers/remote/cluster_cache.go#L332-L397

tl;dr (defaults, but configurable):
- poll every 10 seconds
- using a 5-second request timeout
- unhealthy threshold of 10
- Try to GET /
- If we get 10+ consecutive failures, it's unhealthy

@jmprusi are you working on this now?

**Comment by jmprusi** [comment]
@ncdc yes, I started working on this. Do you think kcp should directly monitor the pcluster? or as we switch to a more agent-based approach, the syncer should be doing that?

There's a better explanation here, of the agent-based approach: https://docs.google.com/document/d/1pDv2OVXpT5TUgO0rKNFwsht1EFasUE_15mvvBY_R-LA/edit#

**Comment by ncdc** [comment]
It's probably better to have the agent heartbeat to kcp, as you've modeled in your doc. There's no guarantee that kcp will be able to connect to a pcluster (firewalls, etc.).

**Comment by pweil-** [comment]
@imjasonh do we consider https://github.com/kcp-dev/kcp/pull/738 as closing this as well as part of the work on pcluster registration?

**Comment by imjasonh** [comment]
> @imjasonh do we consider #738 as closing this as well as part of the work on pcluster registration?

Not completely, though it's progress. We still need the syncer to periodically make the heartbeat update. #738 is only responsible for reacting to the lack of recent-enough heartbeats by marking the cluster as un-`Ready`.
